analyzers/src/SonarAnalyzer.CSharp/Rules/InfiniteRecursion_RoslynCfg.cs

```diff
@@ -0,0 +1,224 @@
+                            cfg = cfg.GetAnonymousFunctionControlFlowGraph(operationWrapper);
+                        }
+                    }
+
+                    cfg = cfg.GetLocalFunctionControlFlowGraph(symbol as IMethodSymbol);
+                }
+
+                var walker = new CfgWalkerForMethod(new RecursionAnalysisContext<ControlFlowGraph>(cfg, symbol, identifier.GetLocation(), c));
+                if (!walker.CheckPaths() && !cfg.ExitBlock.IsReachable)
+                {
+                    c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, identifier.GetLocation(), "method's recursion"));
+                }
+            }
+
+            private class CfgWalkerForMethod : CfgRecursionSearcher
+            {
+                private readonly BooleanBlockMatcher existReferenceToDeclaringSymbolMatcher;
+
+                public CfgWalkerForMethod(RecursionAnalysisContext<ControlFlowGraph> context)
+                    : base(
+                        context.ControlFlowGraph,
+                        context.AnalyzedSymbol,
+                        context.SemanticModel,
+                        () => context.AnalysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.IssueLocation, "method's recursion")))
+                {
+                    existReferenceToDeclaringSymbolMatcher = new BooleanBlockMatcher(operation =>
+                    {
+                        if (!(operation.Instance.Syntax is InvocationExpressionSyntax invocation))
+                        {
+                            return false;
+                        }
+
+                        return IsInstructionOnThisAndMatchesDeclaringSymbol(invocation.Expression, declaringSymbol, semanticModel);
+                    });
+                }
+
+                protected override bool BlockHasReferenceToDeclaringSymbol(BasicBlock block) =>
+                    existReferenceToDeclaringSymbolMatcher.Matches(block);
+            }
+
+            private class CfgWalkerForProperty : CfgRecursionSearcher
+            {
+                private readonly bool isSet;
+                private readonly BooleanBlockMatcher existReferenceToDeclaringSymbolMatcher;
+
+                public CfgWalkerForProperty(RecursionAnalysisContext<ControlFlowGraph> context, string reportOn, bool isSetAccessor)
+                    : base(
+                        context.ControlFlowGraph,
+                        context.AnalyzedSymbol,
+                        context.SemanticModel,
+                        () => context.AnalysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.IssueLocation, reportOn)))
+                {
+                    isSet = isSetAccessor;
+                    existReferenceToDeclaringSymbolMatcher = new BooleanBlockMatcher(operation =>
+                        operation.Instance.Syntax is var syntax
```

Comment:
It makes it easier to identify that it's a lambda arg
```suggestion
                    existReferenceToDeclaringSymbolMatcher = new BooleanBlockMatcher(x =>
                        x.Instance.Syntax is var syntax
```